### PR TITLE
CONTRACTS: Front-end: frees clause improvements

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
Depends on https://github.com/diffblue/cbmc/pull/7086.

The first commit in this PR is from https://github.com/diffblue/cbmc/pull/7086, please do not read.

The second commit adds the following to the front-end:
- add `__CPROVER_freeable_t` built-in type
- add `__CPROVER_freeable` built-in function
- allow calls to `__CPROVER_freeable_t` functions in frees clauses
- add `__CPROVER_is_freeable` built-in predicate
- add `__CPROVER_is_freed` built-in predicate

The predicates are not yet supported in the back-end and behave as nondet functions

The build fails with ubuntu 18.04 and I am not sure why, since no error is visible on the output.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_base.cpp [hunk 1]
```diff
@@ -777,6 +777,42 @@ void c_typecheck_baset::typecheck_declaration(
...
+
+          return to_symbol_expr(expr).get_identifier() == id;
+        };
+
+        if(!has_subexpr(expr, pred))
+          return;
+
+        error().source_location = expr.source_location();
+        error() << id2string(id) + " is not allowed in preconditions." << eom;
+        throw 0;
```
Comment (feliperodri) [inline src/ansi-c/c_typecheck_base.cpp:795]: I guess the previous PR contains this test case, right?

File: src/ansi-c/c_typecheck_code.cpp [hunk 2]
```diff
@@ -845,9 +857,15 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
     typecheck_spec_assigns(
       static_cast<unary_exprt &>(code.add(ID_C_spec_assigns)).op().operands());
   }
+
+  if(code.find(ID_C_spec_frees).is_not_nil())
+  {
+    typecheck_spec_frees(
+      static_cast<unary_exprt &>(code.add(ID_C_spec_frees)).op().operands());
```
Comment (feliperodri) [inline src/ansi-c/c_typecheck_code.cpp:864]: Why no coverage here?

File: src/ansi-c/c_typecheck_code.cpp [hunk 3]
```diff
@@ -950,36 +972,84 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
...
+  else if(can_cast_expr<side_effect_expr_function_callt>(target))
+  {
+    // A call to a freeable_t function symbol without other side effects
+    const auto &funcall = to_side_effect_expr_function_call(target);
+
+    if(!can_cast_expr<symbol_exprt>(funcall.function()))
+    {
+      error().source_location = target.source_location();
+      error() << "function pointer calls not allowed in frees clauses" << eom;
+      throw 0;
```
Comment (feliperodri) [inline src/ansi-c/c_typecheck_code.cpp:1026]: Why no coverage?

File: src/ansi-c/c_typecheck_code.cpp [hunk 4]
```diff
@@ -950,36 +972,84 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
...
+    }
+
+    bool has_freeable_type =
+      (type.id() == ID_empty) &&
+      (type.get(ID_C_typedef) == CPROVER_PREFIX "freeable_t");
+    if(!has_freeable_type)
```
Comment (feliperodri) [inline src/ansi-c/c_typecheck_code.cpp:1032]:
```suggestion
    if((type.id() == ID_empty) &&
       (type.get(ID_C_typedef) == CPROVER_PREFIX "freeable_t"))
```

File: src/ansi-c/c_typecheck_code.cpp [hunk 5]
```diff
@@ -1123,6 +1196,22 @@ void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
   std::swap(targets, tmp);
 }
 
+void c_typecheck_baset::typecheck_spec_frees(exprt::operandst &targets)
+{
+  const std::function<void(exprt &)> typecheck_target = [&](exprt &target) {
+    typecheck_spec_frees_target(target);
+  };
+  typecheck_conditional_targets(targets, typecheck_target, "frees");
```
Comment (feliperodri) [inline src/ansi-c/c_typecheck_code.cpp:1204]: Should _"frees"_  and _"assigns"_ be a macro?

File: src/ansi-c/parser.y [hunk 6]
```diff
@@ -3394,6 +3408,27 @@ cprover_contract_assigns_opt:
...
+        {
+          $$=$1;
+          set($$, ID_C_spec_frees);
+          mto($$, $3);
+        }
+        | TOK_CPROVER_FREES '(' ')'
+        {
+          $$=$1;
+          set($$, ID_C_spec_frees);
+          parser_stack($$).add_to_operands(exprt(ID_target_list));
```
Comment (feliperodri) [inline src/ansi-c/parser.y:3422]: No coverage?

**Review by feliperodri — COMMENTED** [review]
Left a few comments, but I will get another review after merging previous PR (or moving all _frees_ clauses code over here and kill dependency).

File: src/ansi-c/c_typecheck_code.cpp (hunk 4 above)
Comment (remi-delmas-3000) [reply src/ansi-c/c_typecheck_code.cpp:1032]: I like naming what is being checked to make the intent more explicit.

File: src/ansi-c/c_typecheck_code.cpp (hunk 5 above)
Comment (remi-delmas-3000) [reply src/ansi-c/c_typecheck_code.cpp:1204]: what is wrong with std::function ?

**Comment by remi-delmas-3000** [comment]
Superseded by https://github.com/diffblue/cbmc/pull/7091

**Pull request closed without merging**